chain/chain/src/store.rs

```diff
@@ -234,6 +292,58 @@ impl ChainStore {
             }
         }
     }
+
+    pub fn check_blocks_on_same_chain(
+        &mut self,
+        cur_header: &BlockHeader,
+        base_block_hash: &CryptoHash,
+        max_difference_in_height: u64,
+    ) -> bool {
+        // first step: update cache head
+        if self.header_history.is_empty() {
+            self.header_history.push_back(cur_header.clone());
+        }
+        let mut cur_block_hash = cur_header.inner.prev_hash;
+
+        if self.header_history.contains(&cur_header.hash) {
+            self.header_history.update(&cur_header.hash, vec![]);
+        } else {
+            let mut header_list = vec![cur_header.clone()];
+            while !self.header_history.is_empty() {
+                let cur_block_header = if let Ok(header) = self.get_block_header(&cur_block_hash) {
+                    header.clone()
+                } else {
+                    return false;
+                };
+                if self.header_history.contains(&cur_block_header.hash) {
+                    self.header_history.update(&cur_block_header.hash, header_list);
+                    break;
+                }
+                cur_block_hash = cur_block_header.inner.prev_hash;
+                header_list.push(cur_block_header);
+            }
+        }
+
+        // second step: check if `base_block_hash` exists
+        if self.header_history.contains(base_block_hash) {
+            return true;
+        }
+        let num_to_fetch = max_difference_in_height - self.header_history.len() as u64;
```

Comment:
This will crash with underflow if the chain you just updated with was longer than the chain you already had in the history